cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -1189,16 +1211,25 @@ func fakeNode(instance cloudprovider.Instance, reason string) *apiv1.Node {
 
 // PeriodicCleanup performs clean-ups that should be done periodically, e.g.
 // each Autoscaler loop.
-func (csr *ClusterStateRegistry) PeriodicCleanup() {
+func (csr *ClusterStateRegistry) PeriodicCleanup(scaleDownDelayAfterFailure time.Duration, currentTime time.Time) {
 	// Clear the scale-up failures info so they don't accumulate.
-	csr.clearScaleUpFailures()
+	csr.clearScaleUpFailures(scaleDownDelayAfterFailure, currentTime)
 }
 
 // clearScaleUpFailures clears the scale-up failures map.
-func (csr *ClusterStateRegistry) clearScaleUpFailures() {
+func (csr *ClusterStateRegistry) clearScaleUpFailures(scaleDownDelayAfterFailure time.Duration, currentTime time.Time) {
 	csr.Lock()
 	defer csr.Unlock()
-	csr.scaleUpFailures = make(map[string][]ScaleUpFailure)
+	newScaleUpFailures := make(map[string][]ScaleUpFailure)
+
+	for nodeGroupId, scaleUpFailures := range csr.scaleUpFailures {
+		for _, scaleUpFailure := range scaleUpFailures {
+			if scaleUpFailure.Time.Add(scaleDownDelayAfterFailure).After(currentTime) {
+				newScaleUpFailures[nodeGroupId] = append(newScaleUpFailures[nodeGroupId], scaleUpFailure)
```

Comment:
We want to clear only stale scale up failures so that `scaleUpFailures` can be used to evaluate if scale down should happen [here](https://github.com/kubernetes/autoscaler/pull/5729/files#diff-5d0598cafbe97c5bf254de21c9425430e21a392aadbedaccc5d78d0fcf169db4R79).